### PR TITLE
Speed up method patching with identity Hashes

### PR DESCRIPTION
**Pull request by amomchilov**
### Motivation

The `@sigs_that_raised` and `@was_ever_final_names` hashes are keyed by types for which `#==` is the same as `Object#equal?` (`Symbol`s and `Module`s). We can speed these up by using [`Hash#compare_by_identity`](https://ruby-doc.org/3.2.2/Hash.html#method-i-compare_by_identity).

Identity hashes are about 1.3x faster at the moment (in Ruby 3.2.1):

```
Warming up --------------------------------------
 regular Hash            1.520M i/100ms
 compare_by_identity     1.948M i/100ms
Calculating -------------------------------------
 regular Hash            15.122M (± 1.4%) i/s -     76.002M in   5.026956s
 compare_by_identity     19.482M (± 1.2%) i/s -     97.422M in   5.001500s

Comparison:
 compare_by_identity: 19481691.7 i/s
 regular Hash       : 15121932.4 i/s - 1.29x  slower
```

<details><summary><code>benchmark.rb</code></summary>

```ruby
#!/usr/bin/ruby

require 'benchmark/ips'

k = Module.new
hash = { k => true }
hash_by_id = { k => true }.compare_by_identity

Benchmark.ips do |x|
  x.report("regular Hash       ") { hash[k.object_id] }
  x.report("compare_by_identity") { hash_by_id[k]     }
  x.compare!
end
```

</details> 

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by froydnj** [comment]
This is very cool!  Verified the same speedup on 2.7 and 3.1.

File: gems/sorbet-runtime/lib/types/private/methods/_methods.rb [hunk 1]
```diff
@@ -192,21 +192,19 @@ def self._check_final_ancestors(target, target_ancestors, source_method_names, s
   end
 
   def self.add_module_with_final_method(mod, method_name, is_singleton_method)
-    m = is_singleton_method ? mod.singleton_class : mod
-    mid = m.object_id
-    methods = @modules_with_final[mid]
+    methods = @modules_with_final[m]
```
Comment (froydnj) [inline gems/sorbet-runtime/lib/types/private/methods/_methods.rb:195]: Looks like you need to put back the definition of `m` here.

Comment (amomchilov) [reply gems/sorbet-runtime/lib/types/private/methods/_methods.rb]: Good catch. This was a leak from another change I was making (foreshadowing! :eyes)

**Comment by froydnj** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_PBFpggwCjFKUkA
→ https://go/builds/bui_PBFpPBCPtYaKuz
→ https://go/builds/bui_PBFpMuyDRxSkJ8

**Comment by amomchilov** [comment]
Hey @froydnj is this good to merge?

**Comment by froydnj** [comment]
> Hey @froydnj is this good to merge?

This looks OK to me, but testing it is running into some bitrot with our sorbet-runtime test scripts (since it's been a while since we've had to test sorbet-runtime) and I'm not entirely sure how long the issues will take to fix.

**Comment by froydnj** [comment]
Stripe folks can go look at https://go/builds/bui_PBinGt7Tf4jaqW for a build with a single test failure (not this PR's fault).

**Review by froydnj — APPROVED** [review]
Thank you!

**Pull request merged**